### PR TITLE
Add --deterministic to restrict training to deterministic kernels

### PR DESCRIPTION
**Pull request by sw005320**
## What did you change?

Two new training options in `espnet2/tasks/abs_task.py`:

- `--deterministic` (default `false`) — calls `torch.use_deterministic_algorithms(True)`, exports `CUBLAS_WORKSPACE_CONFIG=:4096:8` before any GEMM runs, and forces cudnn-benchmark off / cudnn-deterministic on.
- `--deterministic_warn_only` (default `false`) — downgrades the `RuntimeError` raised for an operation with no deterministic implementation into a warning.

The switch itself lives in `espnet2/torch_utils/set_all_random_seed.py` as `set_deterministic()`, next to `set_all_random_seed()`. Also adds unit tests and a paragraph to `doc/espnet2_training_option.md`.

Both options default to `false`, so nothing changes for existing configs.

---

## Why did you make this change?

Related to #6433, where five runs sharing `seed=0` differ by up to 19 points of CER, with the per-sample CTC losses identical up to the first backward pass.

Seeding is not the gap. `set_all_random_seed()` calls `torch.random.manual_seed()`, which already seeds CUDA, MPS and XPU — so the claim in [this comment](https://github.com/espnet/espnet/issues/6433#issuecomment-4410578269) that CUDA RNG is left unseeded is not accurate, and the reporter's observation that the forward pass is bit-wise reproducible confirms it.

What ESPnet had no switch for is the other half of the [PyTorch randomness notes](https://pytorch.org/docs/stable/notes/randomness.html): CUDA kernels that accumulate with atomics, whose result depends on the order the threads happen to finish in. `--cudnn_deterministic` covers cuDNN only, so gradients diverge from the first backward pass and the difference compounds over training.

`torch.nn.CTCLoss` is the operation this trips over in practice — its native CUDA backward has no deterministic implementation, and cuDNN's is used only when the inputs meet the conditions in the PyTorch document. That is why `--deterministic_warn_only` is offered alongside: it keeps the rest of the model deterministic instead of stopping the run.

---

## Is your PR small enough?

Yes — 4 files, ~120 lines.

---

## Additional Context

- Issue: #6433
- Not covered here: making CTC itself deterministic on CUDA (e.g. routing the builtin CTC through CPU or enforcing the cuDNN preconditions), and the espnet3 / Lightning path, which has `L.seed_everything(seed, workers=True)` but no `Trainer(deterministic=...)`. Happy to follow up in separate PRs if wanted.

### CODE REVIEW

File: espnet2/torch_utils/set_all_random_seed.py [hunk 1]
```diff
@@ -7,4 +9,40 @@
...
+    # NOTE(kamo): cuBLAS reads CUBLAS_WORKSPACE_CONFIG when it creates its
+    # handle, i.e. at the first GEMM call, so this must be set before any
+    # computation happens.
+    if "CUBLAS_WORKSPACE_CONFIG" not in os.environ:
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
+    else:
+        logging.info(
+            "CUBLAS_WORKSPACE_CONFIG is already set to "
+            f"'{os.environ['CUBLAS_WORKSPACE_CONFIG']}' and it is kept as is. "
+            "':4096:8' or ':16:8' is required for deterministic cuBLAS operations."
```
[the remark this answers is not in the document]
Comment (sw005320) [reply espnet2/torch_utils/set_all_random_seed.py]: Valid at the commit you reviewed, and already fixed in `9ef4d52f6` — I should have closed the thread then rather than leaving it open.

At `83608829d` the code did keep whatever was there, and only said so at `info`:

```python
if "CUBLAS_WORKSPACE_CONFIG" not in os.environ:
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
else:
    logging.info(
        "CUBLAS_WORKSPACE_CONFIG is already set to "
        f"'{os.environ['CUBLAS_WORKSPACE_CONFIG']}' and it is kept as is. "
        "':4096:8' or ':16:8' is required for deterministic cuBLAS operations."
    )
torch.use_deterministic_algorithms(True, warn_only=warn_only)
```

So `CUBLAS_WORKSPACE_CONFIG=:2:2` survived and the first GEMM raised, with only an `info` line to explain it.

Current head validates against both accepted values and replaces anything else, warning rather than informing when it does:

```python
DETERMINISTIC_CUBLAS_CONFIGS = (":4096:8", ":16:8")
...
config = os.environ.get("CUBLAS_WORKSPACE_CONFIG")
if config not in DETERMINISTIC_CUBLAS_CONFIGS:
    if config is not None:
        logging.warning(
            f"CUBLAS_WORKSPACE_CONFIG='{config}' is not one of "
            f"{DETERMINISTIC_CUBLAS_CONFIGS}, so cuBLAS operations would "
            "raise a RuntimeError in deterministic mode. Overwriting it "
            f"with '{DETERMINISTIC_CUBLAS_CONFIGS[0]}'"
        )
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = DETERMINISTIC_CUBLAS_CONFIGS[0]
torch.use_deterministic_algorithms(True, warn_only=warn_only)
```

`None` is handled by the same branch, so the unset case still gets `:4096:8` without a spurious warning. `warning` rather than `info` because silently overriding something the user set in their environment is worth seeing.
